math/int.go

```diff
@@ -420,10 +421,90 @@ func (i *Int) Unmarshal(data []byte) error {
 	return nil
 }
 
+var bigMaxUint64 = new(big.Int).SetUint64(uint64(stdmath.MaxUint64))
+
 // Size implements the gogo proto custom type interface.
-func (i *Int) Size() int {
-	bz, _ := i.Marshal()
-	return len(bz)
+// Reduction power of 10 is the smallest power of 10, than 1<<64-1
+//
+//	18446744073709551615
+//
+// and the next value fitting with the digits of (1<<64)-1 is:
+//
+//	10000000000000000000
+var big10Pow19, _ = new(big.Int).SetString("1"+strings.Repeat("0", 19), 10)
+var log10Of2 = stdmath.Log10(2)
+
+func (i *Int) Size() (size int) {
+	sign := i.Sign()
+	if sign == 0 { // It is zero.
+		// log*(0) is undefined hence return early.
+		return 1
+	}
+
+	ii := i.i
+	alreadyMadeCopy := false
+	if sign < 0 { // Negative sign encountered, so consider len("-")
+		// The reason that we make this comparison in here is to
+		// allow checking for negatives exactly once, to reduce
+		// on comparisons inside sizeBigInt, hence we make a copy
+		// of ii and make it absolute having taken note of the sign
+		// already.
+		size += 1
+		// We already accounted for the negative sign above, thus
+		// we can now compute the length of the absolute value.
+		ii = new(big.Int).Abs(ii)
+		alreadyMadeCopy = true
+	}
+
+	// From here on, we are now dealing with non-0, non-negative values.
+	return size + sizeBigInt(ii, alreadyMadeCopy)
+}
+
+func sizeBigInt(i *big.Int, alreadyMadeCopy bool) (size int) {
```

Comment:
I am a bit confused, since this feels like we could solve this with more pre-computation. E.g. for every bitlen, store "markers" in a map for values of that bitlength that corresspond to different sizes.

(And then we'd have 0 alloc's per size call)

Reply:
Nah, it isn't as simple. I had independently tried this approach out, there are a bunch of numbers which are in the boundary of bit lengths but require you to search between the closest values: and roughly I also checked for the allocations reduction and it wasn't significant. I shall refine it in the future, but for purpose of clear wins this approach as is will work great: the approach to get accuracy is to:
* while precomputing the lookup table, for the same bit length B: store []*big.Int{startingBitLenValue, ...endBitLenValue}
* when you lookup by bit length, perform a binary search (range search) in the retrieved list and see if the value falls within that range


This was the code for my prior experiment which I'll refine after this PR lands:

```go
type savings struct {
	bi *big.Int
	bl int
}

var bLenMap map[int]*savings

func computeLUT() {
	// 1. Compute the tables on which we have a divergence in digits
	// for log values.
	// Goal to lookup the number of digits given a bit length
	bLenMap = map[int]*savings{
		0: {new(big.Int).SetUint64(1), 1},
	}
	for i := uint(0); i <= 258; i++ {
		ii := new(big.Int).SetUint64(1)
		ii = ii.Lsh(ii, i)
		len10Digits := len(ii.String())
		if false {
			fmt.Printf("%-79s bitLen: %-4d %d\n", ii, ii.BitLen(), len10Digits)
		}
		bLenMap[ii.BitLen()] = &savings{bl: len10Digits, bi: ii}
	}
}

func init() {
	computeLUT()
}

func lookup(ii *big.Int) int {
	ii = ii.Abs(ii) // Could be lazily computed using sync.Once only when needed
	bits := ii.BitLen()

	closest, ok := bLenMap[bits]
	if !ok {
		return 1
	}

	switch cmp := ii.Cmp(closest.bi); {
	case cmp == 0:
		return closest.bl

	case cmp < 0:
		for {
			bits--
			retr, ok := bLenMap[bits]
			if !ok {
				break
			}

			if ii.Cmp(retr.bi) >= 0 {
				break
			}
			closest = retr
		}

	case cmp > 0:
		for {
			bits++
			retr, ok := bLenMap[bits]
			if !ok {
				break
			}

			if ii.Cmp(retr.bi) <= 0 {
				if retr.bi.BitLen() == ii.BitLen() {
					closest = retr
				}
				break
			}
			closest = retr
		}
	}
	return closest.bl
}
```